### ISSUE TITLE
growpart on new (C5/M5) AWS instances fails to increase root partition

### ISSUE THREAD

**Issue by polynomial** [NixOS/nixpkgs#33092]
## Issue description

When trying to use the new AWS C5 and M5 instance types `growpart` fails with an error in stage-1-init:
```
[    2.942859] stage-1-init: FAILED: partition-number must be a number
```
While investigating I found that these new instances use a different storage driver for the root device:
http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nvme-ebs-volumes.html
This causes issue because the way growpart is called in stage-1-init from here:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/grow-partition.nix#L36-L44
```
    boot.initrd.postDeviceCommands = ''
      rootDevice="${config.fileSystems."/".device}"
      if [ -e "$rootDevice" ]; then
        rootDevice="$(readlink -f "$rootDevice")"
        parentDevice="$(lsblk -npo PKNAME "$rootDevice")"
        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "''${rootDevice#$parentDevice}"
        udevadm settle
      fi
    '';
```
calls it with `partition-number` of `p1` instead of just `1`: (boot.trace enabled dmesg output)
```
[    3.110620] stage-1-init: + TMPDIR=/run sh /nix/store/wv2hhn6xcnfqgd7blq8fp27gg8fvzxqj-extra-utils/bin/growpart /dev/nvme0n1 p1
[    3.111210] stage-1-init: FAILED: partition-number must be a number
```

I found what I believe is a related issue: https://bugs.launchpad.net/cloud-initramfs-tools/+bug/1697695

So that I can keep working on testing out these new instance types I put a probably really gross hack in a fork of nixpkgs:
https://github.com/extole/nixpkgs/pull/3/files#diff-8e04ff5e80705e77d2a4ff12ea3dda78
```
-        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "''${rootDevice#$parentDevice}"
 +        # support nvme style partitions on AWS
 +        if echo "$parentDevice" | grep nvme >/dev/null 2>&1; then
 +          partitionNumber="''${rootDevice#''${parentDevice}p}"
 +        else
 +          partitionNumber="''${rootDevice#$parentDevice}"
 +        fi
 +        TMPDIR=/run sh $(type -P growpart) "$parentDevice" "$partitionNumber"
```

### Steps to reproduce

Build an AMI off of nixpkgs@67f9fd9e9f1b8e6e97e5f495ebf5e58f79c5c318 and deploy the AMI with a root device definition larger than the AMI so it tries to resize the partition on boot. Login to the instance and note that its root partition is the AMI's size not what you gave it on launch.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.
```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.70, NixOS, 18.03pre-git (Impala)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.16`
 - channels(root): `"nixos-18.03pre-git"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`
```

**Comment by greglearns** [comment]
I've run into this same issue today with brand new installs of Nix 2.0 and nixops.

**Comment by copumpkin** [comment]
Thanks @polynomial for the in-depth investigation!

**Comment by ngortheone** [comment]
+1

**Comment by copumpkin** [comment]
I think this is fixed by #39164
